iot/end_to_end_example/device/app/src/main/kotlin/com/example/androidthings/fancontrol/FanControlActivity.kt

```diff
@@ -0,0 +1,379 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+ * Copyright 2018 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.androidthings.fancontrol
+
+import android.app.Activity
+import android.content.Context
+import android.graphics.Color
+
+import android.util.Log
+import com.google.android.things.contrib.driver.apa102.Apa102
+import com.google.android.things.contrib.driver.ht16k33.AlphanumericDisplay
+import com.google.android.things.pio.Gpio
+import com.google.android.things.pio.PeripheralManager
+import java.io.ByteArrayOutputStream
+import java.io.IOException
+import java.io.InputStream
+
+import android.net.ConnectivityManager
+import android.os.Bundle
+import android.os.Handler
+import android.os.Message
+import android.view.KeyEvent
+
+import com.google.android.things.contrib.driver.button.Button
+import com.google.android.things.contrib.driver.button.ButtonInputDriver
+
+import com.google.android.things.iotcore.ConnectionParams
+import com.google.android.things.iotcore.IotCoreClient
+import com.google.android.things.iotcore.OnConfigurationListener
+import com.google.android.things.iotcore.TelemetryEvent
+
+import org.json.JSONException
+import org.json.JSONObject
+
+import java.io.UnsupportedEncodingException
+import java.security.KeyFactory
+import java.security.KeyPair
+import java.security.NoSuchAlgorithmException
+import java.security.spec.InvalidKeySpecException
+import java.security.spec.PKCS8EncodedKeySpec
+import kotlin.text.Charsets.UTF_8
+
+class FanControlActivity : Activity() {
+
+    private var m_fanOn = false
+    private var m_currTemp = 0.0f
+    private val FRAME_DELAY_MS = 100
+
+    private var mButtonInputDriver: ButtonInputDriver? = null
+    private var mDisplay: AlphanumericDisplay? = null
+
+    private var mLedstrip: Apa102? = null
```

Comment:
do we expect some times these are null? or should it always initialize to something? if the latter, use `lateinit`?

Reply:
Good catch. Turns out the display and ledstrip vars are optional, the others are lateinit. This is now cleaned up.